### PR TITLE
feat: added regex support for scopes

### PR DESCRIPTION
**Pull request by puppybits**
RegEx support for scopes. This enables support for any project management system to enforce naming conventions in their PRs. 

```
# You can define a list of valid scopes
scopes:
  - scope1
  - /^TASK-(\\d)+$/
  ...
```

### CODE REVIEW

**Comment by puppybits** [comment]
@zeke would you be ok with allowing regular expression support? 

As an engineering manager it would help a ton to ensure all PRs are linked to our project management app.

File: lib/is-semantic-message.js [hunk 1]
```diff
@@ -17,6 +17,8 @@ module.exports = function isSemanticMessage (message, validScopes, validTypes, a
 
   const [result] = commits
   const { scope, type } = result.header
-  const isScopeValid = !validScopes || !scope || validScopes.includes(scope)
+  const isScopeValid = !validScopes || !scope || validScopes.some(function(valid) { 
+    return new RegExp(valid.replace(/(^\/|\/$)/g, "")).test(scope)
```
Comment (zeke) [inline lib/is-semantic-message.js:21]: Curious if there's a way to support this without requiring this `replace` call, as it makes some assumptions about the format of the incoming regex...

**Review by zeke — COMMENTED** [review]
Hey thanks for the PR. I'd be happy to accept this change,  but can you add a few more tests for different variants of the scope regexes? This might even be a good time to move that code into its own module with its own tests so we don't have to duplicate the whole setup of handle-pull-request-change tests. Does that make sense?

File: lib/is-semantic-message.js (hunk 1 above)
Comment (puppybits) [reply lib/is-semantic-message.js:21]: I see your point. The thought was to differentiate in the yaml between a regex and a vanilla string. 

Instead the scope could just accept and run everything as a regex test. All of these would be valid scopes: 
```yaml
# You can define a list of valid scopes
scopes:
  - scope1
  - ^TASK-(\\d)+$
  - scope\\d
  ...
```

Then the function would work like this:
```ts
const isScopeValid = !validScopes || !scope || validScopes.some(function(valid) { 
    return new RegExp(valid).test(scope) 
}
```

Would that be better?

**Comment by puppybits** [comment]
> Hey thanks for the PR. I'd be happy to accept this change, but can you add a few more tests for different variants of the scope regexes? This might even be a good time to move that code into its own module with its own tests so we don't have to duplicate the whole setup of handle-pull-request-change tests. Does that make sense?

Sure. Thanks.

File: lib/is-semantic-message.js (hunk 1 above)
Comment (puppybits) [reply lib/is-semantic-message.js]: We need some way to know to match as a regex or an exact match. 

The issue is if everything ran as a regex then we'd be accepting substring.
`new RegExp('validScope').test('invalidScope') => true (but should fail)`
`'validScope' === 'invalidScope'  => false (which is correct)`

I assumed if the scope starts & ends as `/` that's would be a good way to know if we run as an exact match or a regex. Either `scope.replace(isRegExp)` or `scope[0] === "/" && scope[scope.length-1] === "/"` are valid test but we would need to remove the first & last slash before we send to the RegExp function.

**Comment by puppybits** [comment]
@zeke I moved all scope checks to an `is-scope-valid.js` file and added a lot more tests. Let me know what you think.

**Comment by zeke** [comment]
Looks great. Thanks for adding all those tests and for preserving existing functionality. ⭐ 

This will be deployed automatically within a few minutes.

**Comment by rittme** [comment]
Maybe it's a coincidence, but the app looks broken since this PR was deployed.
https://github.com/zeke/semantic-pull-requests/issues/125

File: lib/is-semantic-message.js [hunk 2]
```diff
@@ -1,22 +1,30 @@
...
+    if (process.env.NODE_ENV !== "test") console.error(error);
+    return false;
   }
 
-  const [result] = commits
-  const { scope: scopes, type } = result.header
-  const isScopeValid = !validScopes || !scopes || scopes.split(',').map(scope => scope.trim()).every(scope => validScopes.includes(scope))
-  return (validTypes || commitTypes).includes(type) && isScopeValid
-}
+  const { type } = result.header;
```
Comment (cenobitedk) [inline lib/is-semantic-message.js:25]: `result` is undefined since the line `const [result] = commits;` has been removed.

**Pull request merged**